### PR TITLE
Add support for SFR Identity authentication

### PR DESCRIPTION
**Pull request by macournoyer**
### What issue does this pull request address?

With Identity authentication coming soon to SFR, we need to use proper OAuth subject ID depending on the rendering backend.

### What is the solution

~~I borrowed the logic from Ubercorn to detect which rendering backend is used: https://github.com/Shopify/ubercorn/blob/master/app/src/background/components/set_render_backend.coffee.~~ (Not true anymore, see e6f5b31)

Then, based on the detected backend, use the proper subject ID.

### What should the reviewer focus on and are there any special considerations?

I might be wrong, but I don't think this requires any changes to the local storage caching.

### CODE REVIEW

File: src/env.ts
Comment (t-kelly) [inline src/env.ts:8]: Is this replaced by `DEV_OAUTH2_SUBJECT_NAME`?

File: src/background.ts
Comment (t-kelly) [inline src/background.ts:20]: It seems like the pattern used to structure these constants doesn't work anymore -- can we think of a cleaner way then assigning nulls?

File: src/background.ts
Comment (t-kelly) [inline src/background.ts:26]: Where is `DEVTOOLS_SCOPE` being defined now? Looks like it was removed in L4?

File: src/background.ts
Comment (t-kelly) [inline src/background.ts:56]: I don't think this check is needed since this extension is only targetting Chrome and not other browsers? If it is needed, we need to do it everywhere else

File: src/manifest.json
Comment (t-kelly) [inline src/manifest.json:6]: Will adding this permission increase review times? I know we had this problem with other permissions, and why `activeTab` is preferred..

Instead of reading request headers, we could be including SFR status in Server Timing header which can be accessed via JS.

File: src/background.ts
Comment (macournoyer) [reply src/background.ts:56]: It was to prevent an error when loading the extension in the Extensions page for Chrome. `chrome.webRequest` is not available on that page. Maybe I should check for `typeof chrome.webRequest !== 'undefined'` instead? Would make this clearer.

File: src/manifest.json
Comment (macournoyer) [reply src/manifest.json:6]: Ah good point! I just copied the way Ubercorn did it. But it seemed a bit fragile (eg.: if you reload the plugin after the page, doesn't work).

Did you have a metric in mind, or just a bogus one?

File: src/background.ts
Comment (macournoyer) [reply src/background.ts:26]: It's a map now: https://github.com/Shopify/shopify-theme-inspector/pull/63/files#diff-9bcf0275d72a0ae3b1d14282a4a1a2dbR20.

One for each backend. Because Identity requires that we use a scope owned by the audience app (Core or SFR).

File: src/env.ts
Comment (macournoyer) [reply src/env.ts:8]: Yes. It now uses [dynamic registration](https://identity.docs.shopify.io/oauth/dynamic-registration-endpoint/) to the the client IDs. So that we don't have to hardcode UUID that depend on fixtures.

It was the approach recommended to me when I tried adding some seeds in the Identity Rails app.

File: src/manifest.json
Comment (t-kelly) [reply src/manifest.json:6]: See my most recent comment, but I don't think we'll need this permission 🤞

File: src/utils/helpers.ts
Comment (macournoyer) [reply src/utils/helpers.ts:45]: Seems `appName` is the same both on SFR and on Core: https://github.com/Shopify/shopify/blob/master/components/merchant_marketing/eagerlib/analytics.rb#L133, https://github.com/Shopify/storefront-renderer/blob/master/app/helpers/content_for_header/analytics.rb#L84

Comment (macournoyer) [reply src/utils/helpers.ts:45]: Ah `window.BOOMR.application` it is!

File: src/background.ts
Comment (macournoyer) [reply src/background.ts]: Decided to switch to using names instead of hardcoding ID of the subjects for both dev and prod. It's one more request, but since the result is cached, I don't think it maters much...

File: src/background.ts
Comment (t-kelly) [inline src/background.ts:9]: On line 122 you call `getOauth2Client()` and `getRenderBackend()`, and then `getRenderBackend()` gets called again here. Is this an expensive operation? Can you see an easy way of eliminating the duplicate call?

**Review by t-kelly — COMMENTED** [review]
This is looking great! Haven't 🎩 'd it, but the code tidying up nicely 😄

**Review by t-kelly — APPROVED** [review]
Code looks good. What's left that blocking you from shipping? Has anyone else been able to test if it works?

**Comment by ADTC** [comment]
Hi, I'd suggest you to pin this issue, and update your front page readme. Seems like you're getting new issues that merely track back to this (including my own 😄).

I'd love to send a PR for the readme update (no code changes).

**Comment by macournoyer** [comment]
Finally found a way to unblock this in the backend I think. Should have things all working this week if all goes well.

**Comment by macournoyer** [comment]
Shop staff & Shopify employees should now be able to profile all shops with this.

Only missing piece is collaborators won't be able to profile on SFR yet. I'm working on it. Pretty sure it won't require any changes to the Theme Inspector. So I think we're good to publish on Chrome Store after this is merged.

Also of note: Theme Inspector Identity/OAuth config can now be changed via this file: https://github.com/Shopify/storefront-renderer/blob/master/config/deploy/production/identity.yml.erb#L27-L64

**Comment by dan-gamble** [comment]
Yayy 👏. I've been keeping a close eye on this.

Would you be able to share a rough ETA on the collab access once this does get merged? I know that was a massive pain point right at the start with this so having that timeline be open would be incredibly helpful

**Comment by macournoyer** [comment]
@dan-gamble technically should only be a few days. I'm working on deploying the fix today. Will update this ticket when it's available.

**Review by wizardlyhel — APPROVED** [review]
🎩  - Works on production SFR for shop owner/staff with permission, and Shopify employees

Just missing a few changes:

* [ ] Update `manifest.json` version to `1.0.8`
* [ ] Update change log

**Comment by wizardlyhel** [comment]
[Non-Blocking] It would be nice if the extension shows which rendering engine it obtain the liquid profile from

**Comment by macournoyer** [comment]
> [Non-Blocking] It would be nice if the extension shows which rendering engine it obtain the liquid profile from

Isn't this a bit redundant if we already have this info in Ubercorn?

**Comment by wizardlyhel** [comment]
> Isn't this a bit redundant if we already have this info in Ubercorn?

For developers that are not Shopify employees :)

There is a huge discrepancy in timing between a core and a SFR render. It will be informative for them to know if the profiling happened on which rendering engine since not all pages are being rendered by SFR as of this current moment.

**Comment by macournoyer** [comment]
Here it is:

Screen Shot 2020-10-02 at 1 28 27 PM

**Comment by wizardlyhel** [comment]
🎩 🎩  On the new addition - Thank you!

File: src/manifest.json
Comment (t-kelly) [inline src/manifest.json:3]: We're adding a feature of knowing which rendering engine the store is on, does that classify a minor version bump? i.e `1.1.0`?

**Review by t-kelly — APPROVED** [review]

**Comment by macournoyer** [comment]
Already been accepted! 1.1.0 is on the Store: https://chrome.google.com/webstore/detail/shopify-theme-inspector-f/fndnankcflemoafdeboboehphmiijkgp?hl=en

**Pull request merged**
